src/Common/ThreadPool.h

```diff
@@ -237,6 +241,15 @@ class ThreadFromGlobalPoolImpl : boost::noncopyable
     {
         if (!state)
             return false;
+        /// The problem is that the our ThreadFromGlobalPool can be actually finished
+        /// before we try to join the thread or check whether it is joinable or not.
+        /// In some places we have code like:
+        ///     if (thread->joinable())
+        ///         thread->join();
+        /// Where join() won't be executed in case when we call it
+        /// from the same std::thread and it will end to std::abort().
+        if (state->finished)
```

Comment:
But joining the thread (std::thread, pthread) from itself is UB, in reality it will likely hang.
Maybe it is better to keep this function consistent with std::thread and leave it as-is.
And instead implicitly reset the state if it is finished in the dtor?

Reply:
Ok. I also had an idea to make this abstraction similar to std::jthread (auto joinable in destructor)